crates/swc_ecma_utils/src/function/fn_env_hoister.rs

```diff
@@ -126,57 +145,285 @@ impl FnEnvState {
             )
         }
     }
-}
 
-impl Take for FnEnvState {
-    fn dummy() -> Self {
-        Self::default()
+    fn get_this(&mut self) -> Ident {
+        self.this
+            .get_or_insert_with(|| private_ident!("_this"))
+            .clone()
     }
-}
 
-pub struct FnEnvHoister<'a> {
-    state: &'a mut FnEnvState,
+    fn super_get(&mut self, id: &mut Ident, span: Span) -> Ident {
+        if let Some(callee) = self.super_get.ident.get(&id.sym) {
+            callee.clone()
+        } else {
+            let ident = private_ident!(span, format!("_superprop_get_{}", id.sym));
+            self.super_get.ident.insert(id.take().sym, ident.clone());
+            ident
+        }
+    }
+
+    fn super_get_computed(&mut self, span: Span) -> Ident {
+        self.super_get
+            .computed
+            .get_or_insert_with(|| private_ident!(span, "_superprop_get"))
+            .clone()
+    }
 }
 
-impl<'a> FnEnvHoister<'a> {
-    pub fn new(state: &'a mut FnEnvState) -> Self {
-        FnEnvHoister { state }
+impl Take for FnEnvHoister {
+    fn dummy() -> Self {
+        Self::default()
     }
 }
 
-impl<'a> VisitMut for FnEnvHoister<'a> {
+impl VisitMut for FnEnvHoister {
     noop_visit_mut_type!();
 
     fn visit_mut_expr(&mut self, e: &mut Expr) {
         match e {
             Expr::Ident(id) if id.sym == js_word!("arguments") => {
                 let arguments = self
-                    .state
                     .args
                     .get_or_insert_with(|| private_ident!("_arguments"));
                 *e = Expr::Ident(arguments.clone());
             }
             Expr::This(..) => {
-                let this = self
-                    .state
-                    .this
-                    .get_or_insert_with(|| private_ident!("_this"));
-                *e = Expr::Ident(this.clone());
+                let this = self.get_this();
+                *e = Expr::Ident(this);
             }
             Expr::MetaProp(MetaPropExpr {
                 kind: MetaPropKind::NewTarget,
                 ..
             }) => {
                 let target = self
-                    .state
                     .new_target
                     .get_or_insert_with(|| private_ident!("_newtarget"));
                 *e = Expr::Ident(target.clone());
             }
+            // super.foo = 123 => super_get_foo = (value) => super.foo = value
```

Comment:
Doesn't we need special handling for `super.foo++`?

Reply:
Yes, I somehow forget that